### PR TITLE
feat: add_zhipuai_model

### PR DESCRIPTION
**Pull request by fengju0213**
## Description

add zhipiai_model

## Motivation and Context

Reduce the threshold for domestic users to use camel

- [ ] I have raised an issue to propose this change ([required](https://github.com/camel-ai/camel/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [x] Example (update in the folder of example)

## Implemented Tasks

- [ ] Subtask 1
- [ ] Subtask 2
- [ ] Subtask 3

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://github.com/camel-ai/camel/blob/master/CONTRIBUTING.md) guide. (**required**)
- [x] My change requires a change to the documentation.
- [] I have updated the tests accordingly. (*required for a bug fix or a new feature*)
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: camel/configs/__init__.py [hunk 1]
```diff
@@ -19,10 +19,16 @@
     OpenSourceConfig,
 )
 
+from .zhipuai_config import (
+    ZHIPU_API_PARAMS,
+    ZhipuaiConfig,)
```
Comment (Wendong-Fan) [inline camel/configs/__init__.py:24]:
```suggestion
    ZhiPuConfig,)
```

File: camel/configs/__init__.py [hunk 2]
```diff
@@ -19,10 +19,16 @@
...
+
 __all__ = [
     'BaseConfig',
     'ChatGPTConfig',
+    'ZhipuaiConfig',
```
Comment (Wendong-Fan) [inline camel/configs/__init__.py:29]:
```suggestion
    'ZhiPuConfig',
```

File: camel/configs/zhipuai_config.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    request_id: Optional[str] = None
+    do_sample: Optional[bool] = True
+    stream: Optional[bool] = False
+    temperature: Optional[float] = 0.95
+    top_p: Optional[float] = 0.7
+    max_tokens: Optional[int] = 8192
+    stop: Optional[List[str]] = None
+    tools: Optional[List[str]] = None
+    tool_choice: Optional[str] = "auto"
+
```
Comment (Wendong-Fan) [inline camel/configs/zhipuai_config.py:74]: content here is not matching with docstring

File: camel/models/zhipuai_model.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+import os
+from typing import Any, Dict, List, Optional, Union
+from camel.messages import OpenAIMessage
+from zhipuai import ZhipuAI
+from zhipuai.core._sse_client import StreamResponse
+from camel.configs import ZHIPU_API_PARAMS
+from camel.models import BaseModelBackend
+from camel.types import Completion, ChatCompletionChunk, ModelType
+from camel.utils import BaseTokenCounter, api_key_required, ZhipuAITokenCounter
+class ZhipuAIModel(BaseModelBackend):
```
Comment (Wendong-Fan) [inline camel/models/zhipuai_model.py:24]:
```suggestion
from camel.utils import BaseTokenCounter, api_key_required, ZhipuAITokenCounter

class ZhipuAIModel(BaseModelBackend):
```

File: camel/models/zhipuai_model.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    r"""ZhipuAI API in a unified BaseModelBackend interface."""
+
+    def __init__(
+        self, model_type: ModelType, model_config_dict: Dict[str, Any]
```
Comment (Wendong-Fan) [inline camel/models/zhipuai_model.py:28]: add `api_key` as argument

File: camel/models/zhipuai_model.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+        super().__init__(model_type, model_config_dict)
+        url = os.environ.get('ZHIPUAI_API_BASE_URL', None)
+        api_key = os.environ.get('ZHIPUAI_API_KEY', None)
+        self._client = ZhipuAI(api_key=api_key,base_url=url)
+        self._token_counter: Optional[BaseTokenCounter] = None
+
+    @api_key_required
+    def run(
+        self,
+        messages:  List[OpenAIMessage],
```
Comment (Wendong-Fan) [inline camel/models/zhipuai_model.py:47]:
```suggestion
        messages: List[OpenAIMessage],
```

File: camel/models/zhipuai_model.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    ) -> Union[Completion, StreamResponse[ChatCompletionChunk]]:
+        r"""Runs inference of ZhipuAI chat completion.
+
+        Args:
+            messages (List[OpenAIMessage]): Message list with the chat history
+                in ZhipuAI API format.
```
Comment (Wendong-Fan) [inline camel/models/zhipuai_model.py:53]:
```suggestion
                in OpenAI API format.
```

File: camel/models/zhipuai_model.py [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+        url = os.environ.get('ZHIPUAI_API_BASE_URL', None)
+        api_key = os.environ.get('ZHIPUAI_API_KEY', None)
+        self._client = ZhipuAI(api_key=api_key,base_url=url)
+        self._token_counter: Optional[BaseTokenCounter] = None
+
+    @api_key_required
+    def run(
+        self,
+        messages:  List[OpenAIMessage],
+    ) -> Union[Completion, StreamResponse[ChatCompletionChunk]]:
```
Comment (Wendong-Fan) [inline camel/models/zhipuai_model.py:48]: the response from Zhipu is different with OpenAI

```suggestion
    ) -> Union[Completion, StreamResponse[ChatCompletionChunk]]:
```

**Review by Wendong-Fan — COMMENTED** [review]
Thanks for the contribution and sorry for the late review, considering the response from Zhipu original SDK is different with OpenAI, I suggest using the OpenAI interface which is supported by Zhipu to unify the response format, refer to https://open.bigmodel.cn/dev/api#sdk_auth

**Comment by fengju0213** [comment]
> Thanks for the contribution and sorry for the late review, considering the response from Zhipu original SDK is different with OpenAI, I suggest using the OpenAI interface which is supported by Zhipu to unify the response format, refer to https://open.bigmodel.cn/dev/api#sdk_auth
I have considered using the interface of OpenAI before, but the parameters of Zhipu cannot be used in this way. Should we convert the response format or directly use the interface of OpenAI？

File: camel/models/zhipuai_model.py (hunk 8 above)
Comment (fengju0213) [reply camel/models/zhipuai_model.py:48]: Should we convert the response format or directly use the interface of OpenAI？

File: camel/models/zhipuai_model.py [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+# distributed under the License is distributed on an “AS IS” BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+
+import os
+from typing import Any, Dict, List, Optional, Union
+from camel.messages import OpenAIMessage
+from zhipuai import ZhipuAI
```
[the remark this answers is not in the document]
Comment (fengju0213) [reply camel/models/zhipuai_model.py:18]: I am considering doing it this way.The advantage of doing so is that it is more convenient , and the disadvantage is that some parameters of zhipu cannot be used

[the remark this answers is not in the document]
Comment (zechengz) [reply camel/models/zhipuai_model.py]: > the disadvantage is that some parameters of zhipu cannot be used

Can you give an example?

[the remark this answers is not in the document]
Comment (fengju0213) [reply camel/models/zhipuai_model.py]: like 'do_sample' ,but it's an unimportant function,so we decide to use OpenAI SDK in last dev meeting

File: examples/zhipuai_models/role_playing_with_zhipuai_model.py
Comment (Wendong-Fan) [inline examples/zhipuai_models/role_playing_with_zhipuai_model.py:1]: for the example file could you refer to our new format here: https://github.com/camel-ai/camel/pull/585/files#diff-e9109504d46cb1cdc90fddd1ec377fc7882ee1f1ec3a17053616ed3765ce0304

don't put the core code within one function, also add the output from your pirnt

**Review by Wendong-Fan — COMMENTED** [review]
Left some comment for small fix~

File: camel/models/zhipuai_model.py [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+        url: Optional[str] = None,
+    ) -> None:
+        r"""Constructor for ZhipuAI backend.
+
+        Args:
+            model_type (ModelType): Model for which a backend is created,
+                such as GLM_* series.
+            model_config_dict (Dict[str, Any]): A dictionary that will
+                be fed into openai.ChatCompletion.create().
+             api_key (Optional[str]): The API key for authenticating with the
```
Comment (Wendong-Fan) [inline camel/models/zhipuai_model.py:48]:
```suggestion
            api_key (Optional[str]): The API key for authenticating with the
```

File: camel/models/zhipuai_model.py [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+        Args:
+            messages (List[OpenAIMessage]): Message list with the chat history
+                in OpenAI API format.
+
+        Returns:
+            Union[ChatCompletion, Stream[ChatCompletionChunk]]:
+                `ChatCompletion` in the non-stream mode, or
+                `Stream[ChatCompletionChunk]` in the stream mode.
+        """
+        response = self._client.chat.completions.create(
```
Comment (Wendong-Fan) [inline camel/models/zhipuai_model.py:78]:
```suggestion
        """
        # Use OpenAI cilent as interface call ZhipuAI
        # Reference: https://open.bigmodel.cn/dev/api#openai_sdk
        response = self._client.chat.completions.create(
```

File: camel/models/zhipuai_model.py [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+                )
+        pass
+
+    @property
+    def stream(self) -> bool:
+        r"""Returns whether the model is in stream mode,
+            which sends partial results each time.
+        Returns:
+            bool: Whether the model is in stream mode.
+        """
```
Comment (Wendong-Fan) [inline camel/models/zhipuai_model.py:121]:
```suggestion
        r"""Returns whether the model is in stream mode, which sends partial
        results each time.

        Returns:
            bool: Whether the model is in stream mode.
        """
```

**Review by Wendong-Fan — APPROVED** [review]

**Pull request merged**